src/server/graceful.rs

```diff
@@ -0,0 +1,620 @@
+            future,
+            state: Some(self.state.clone()),
+        }
+    }
+}
+
+struct GracefulState {
+    counter: AtomicUsize,
+    waker_list: Mutex<Slab<Option<Waker>>>,
+}
+
+impl GracefulState {
+    fn subscribe(&self) {
+        self.counter
+            .fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+    }
+
+    fn unsubscribe(&self) {
+        if self
+            .counter
+            .fetch_sub(1, std::sync::atomic::Ordering::SeqCst)
+            == 1
+        {
+            let mut waker_list = self.waker_list.lock().unwrap();
+            for (_, waker) in waker_list.iter_mut() {
+                if let Some(waker) = waker.take() {
+                    waker.wake();
+                }
+            }
+        }
+    }
+}
+
+pin_project! {
+    /// A wrapper around a future that's being watched for graceful shutdown.
+    ///
+    /// This is returned by [`GracefulShutdown::watch`].
+    ///
+    /// # Panics
+    ///
+    /// This future might panic if it is polled
+    /// after the internal future has already returned `Poll::Ready` before.
+    ///
+    /// Whether or not this future panics in such cases is an implementation detail
+    /// and should not be relied upon.
+    pub struct GracefulFuture<C> {
+        #[pin]
+        future: GracefulConnectionFuture<C, GracefulWaiter>,
+        state: Option<Arc<GracefulState>>,
+    }
+}
+
+impl<C> Debug for GracefulFuture<C> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("GracefulFuture").finish()
+    }
+}
+
+impl<C> Future for GracefulFuture<C>
+where
+    C: GracefulConnection,
+{
+    type Output = C::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        match this.future.poll(cx) {
+            Poll::Ready(v) => {
+                this.state.take().unwrap().unsubscribe();
+                Poll::Ready(v)
+            }
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+/// A future that blocks until a graceful shutdown is complete.
+pub struct GracefulWaiter {
+    state: GracefulWaiterState,
+}
+
+impl GracefulWaiter {
+    fn new(state: Arc<GracefulState>) -> Self {
+        Self {
+            state: GracefulWaiterState { state, key: None },
+        }
+    }
+}
+
+impl Future for GracefulWaiter {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut task::Context<'_>) -> Poll<Self::Output> {
+        let state = &mut self.state;
+
+        if state
+            .state
+            .counter
+            .load(std::sync::atomic::Ordering::SeqCst)
+            == 0
+        {
+            return Poll::Ready(());
+        }
+
+        let mut waker_list = state.state.waker_list.lock().unwrap();
+
+        if state
+            .state
+            .counter
+            .load(std::sync::atomic::Ordering::SeqCst)
+            == 0
+        {
+            // check again in case of race condition
```

Comment:
What's the scenario for race condition here?

Reply:
We first check the counter. That one is zero only if shutdown is already triggered. We then wait for a lock. However it is possible that that we had to wait for the lock while something else was holding the lock to awake the ones waiting for shutdown.

If we do not check again once we have the lock it is possible that the counter in meanwhile is zero. If so we either way need to trigger shutdown here or else we might not be woken up.

This was discovered while developing the origin code on which this PR was inspired (https://github.com/plabayo/tokio-graceful, a crate of mine), where loom helped me to find that issue.